### PR TITLE
Accept new connections in place of old ones when appropriate #5224

### PR DESCRIPTION
**Pull request by MikolajTwarog**
As suggested I added replacing connection when the current one is older than 5 seconds. To do that I added new field in Statistics structure.

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
I am not sure this is how I imagine this to work. My personal thought on how I would handle this would be to add a mechanism to actively check the liveness of the connection, and only drop it if we deem it to be "dead". In this specific case we might drop a TCP connection over a UDP one which has very latient packets.

**Comment by calmh** [comment]
I would settle for accepting a new connection if the old connection-to-be-replaced is at 10s old or something like that. That would add as a tie breaker in simultaneous connections, so we don't end up with dropping both of them, and it would handle replacing dead connections that only the other side has realized is dead.

In theory we might replace a connection with something worse (tcp -> quic), but that would only happen if the other side doesn't have the tcp connection any more and the quic connection is the attempt that goes through. We'll continue trying higher priority protocols so if tcp works we'll get back to it in time.

**Comment by AudriusButkevicius** [comment]
Fine I guess. Do we really want the start time to live on the stats part?

**Comment by MikolajTwarog** [comment]
I thought it would be better than adding new method to Connection.

**Comment by MikolajTwarog** [comment]
Should I change anything more?

**Comment by AudriusButkevicius** [comment]
Idk, it's probably fine, just want @calmh to confirm.

**Comment by imsodin** [comment]
I think this is a good step.  
Ideally considering the huge rat-tail caused by a connection change, I'd issue a ping immediately on an incoming connection attempt. and if that fails drop the old connection and add this connection (and do this ping close dance outside of the `handle` loop and pass it back in if appropriate). If you feel like trying something like that (no guarantees it actually works out), that would of course be great, but I am also fine with the mechanism in this PR.

**Comment by MikolajTwarog** [comment]
Right now I don't have enough time to try this approach. Maybe it would be good to create a new issue about this, so I can work on in some time later.

**Comment by AudriusButkevicius** [comment]
I'll leave for @calmh to merge I guess.

File: lib/connections/service.go [hunk 1]
```diff
@@ -276,7 +276,7 @@ func (s *service) handle(ctx context.Context) {
 		ct, connected := s.model.Connection(remoteID)
 
 		// Lower priority is better, just like nice etc.
-		if connected && ct.Priority() > c.priority {
+		if connected && (ct.Priority() > c.priority || time.Since(ct.Statistics().StartedAt) > 10*time.Second) {
```
Comment (calmh) [inline lib/connections/service.go:279]: This would be a good place for a named constant, `minConnectionReplaceAge`?

Comment (MikolajTwarog) [reply lib/connections/service.go]: I think you are right. I added it.

**Comment by calmh** [comment]
Does it work though? :) I mean, it looks like it should, but did/can we test this somehow?

**Comment by MikolajTwarog** [comment]
I was testing it using my two devices. I was switching WiFi on my second device. Every time I tried it new connection was accepted instead of the old one. It was not the case before my changes.

**Comment by GermanCoding** [comment]
By the way there was a TODO in the code (https://github.com/syncthing/syncthing/pull/6548/files#diff-d31f92d68068bfede5c6b6b01e1b558bL282-L287) that still exists altough it was presumably fixed by this PR.

**Pull request merged**